Introduce semantic conventions for CloudEvents (#16)
dynatrace-oss-contrib/opentelemetry-specification: specification/trace/semantic_conventions/cloudevents.md

# CloudEvents

This document defines how to describe sending and receiving [CloudEvents](https://cloudevents.io/) with spans.

> The transport involved in sending/receiving CloudEvents is represented by a separated span, depending on the underlying protocol used.

Reviewer: ```suggestion
The transport involved in sending/receiving CloudEvents is represented by a separate span, depending on the underlying protocol used.
```

# CloudEvents

> This document defines how to apply semantic conventions when instrumenting code that uses [CloudEvents](https://cloudevents.io/).

Reviewer 2: "uses" is to unclear here IMHO, especially for the emitter side. Should the span encompass the whole operation triggering the event (maybe using the attributes defined here just as additional attributes to the "main" span attributes, e.g. supplementing DB conventions for a SQL INSERT operation) or just the emitting itself? The latter is how it sounds currently, but I think the former would also be interesting.

Author: It's the latter. Since the whole messaging space is still a WIP I don't think we can even encompass a "whole operation of triggering an event", since the event can "flow" into many parts, and it is still unclear how propagation will happen in those cases.

The goal was only to offer attributes that can be added to spans that, for example, are manually created by users when they send cloudevents. Does that makes sense?

Reviewer 2: > Since the whole messaging space is still a WIP

I thought this is no longer related to the messaging spec?

> The goal was only to offer attributes that can be added to spans that, for example, are manually created by users when they send cloudevents. Does that makes sense?

That's the case for all semantic conventions.

> I don't think we can even encompass a "whole operation of triggering an event", since the event can "flow" into many parts

It's true that for HTTP we discussed that having a separate parent per outgoing request makes sense. But still, it would be useful to somehow know which operation was the logical cause of the CloudEvent, assuming that a span for that operation exists.

Author: I re-worded this portion a bit in an attempt to clarify that we are only talking about sending/receiving events and not all that can come along with it. Hopefully it's clearer now. I agree with the points you made, but I feel that this whole topic is really not mature enough for us to come up with anything concrete. At least having the attributes which we know are clear can help bring interest and kick start discussions and ideas.

> These semantic conventions are general and not tied to any protocol used to send/receive the events.

Reviewer 2: "General" conventions are conventions that should be used with another "main" type of span and thus usually do not prescribe a name or kind. I don't think this is what you meant.
I would leave out this sentence entirely, as freestanding semantic conventions are more or less the default anyway.

Reviewer: This could be reworded to explicitly point out this is only about the event (payload) and that transport would be represented by a separate span depending on the underlying protocol used.

Author: Good point, I also don't like the way it is. Will try to come up with something close to what @arminru  mentioned. I think it's valid to emphasize that this is not related to the transport.

Author: Reworked on this part, hopefully now it's clear that it's only talking about sending/receiving the events.

**Status**: [Experimental](../../document-status.md)

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Span name](#span-name)
- [Span kind](#span-kind)
- [Attributes](#attributes)

<!-- tocstop -->

## Span name

> The span name SHOULD be set to the `cloudevents.` prefix, followed by the event type in the following format:

Reviewer 2: We typically don't use a "type" prefix. So using just the event type name sounds more consistent with our other semantic conventions.

Reviewer: That's true but I wonder if it wouldn't make more sense to have the "type" in the span name so once can tell right from looking at the span what this is about without having to find the most specific span attributes being present. It would also make clear according to which semconv the span name is composed. Either way, nothing that would fit into this PR.
I think you could leave the cloudevents.* in though, we have the same thing in the HTTP semconv, for example.

Author: >can tell right from looking at the span what this is about without having to find the most specific span attributes being present

That was exactly my reasoning. In a myriad of spans I think having these with `cloudevents` help distinguish them. Also had the same thought that the HTTP spans also have this "prefix" more or less.

Reviewer 2: For HTTP, yes but only as a fallback if you use just the HTTP verb (since just GET would be a unhelpful span name).
With cloud events, I think the event name will usually make a good span name, even without a cloudevent prefix. But if we only trace the emitting itself not the operation that the event is about then using a prefix would make sense (though then I would use `CloudEvent ` as prefix not `cloudevent.`)

Author: I modified the instructions to use `CloudEvents {event type}` as the span name.

```

> cloudevents.<event type name>

Author: I had initially the span name as `cloudevents.<event type> <operation name>` where the `operation name` followed the ones listed for messaging systems (`send`, `receive` and `process`). But I felt that adding it here now that this is completely separated from messaging would cause confusion so I left it out. What do you think? 

If I add it here, then I also need to list the operation names and it will be somewhat duplicated from the `messaging.md` file. Since they made it very clear from the PR discussions and the SIG meeting that CloudEvents are NOT a messaging system I feel that adding this will trigger again the discussions above, even though in my view as a user it makes sense.

Reviewer: If we don't want to distinguish receive and process then the span kind will be sufficient to tell which end of the operation is represented here.

Author: Yeah I also think so. With the span kind it should be sufficient to tell which end it is.

```

The event type name SHOULD only be used for the span name if it is known to be of low cardinality (cf. [general span name guidelines](../api.md#span)).

Examples:

* `cloudevents.com.github.pull_request.opened`

## Span kind

> A producer of an event should set the span kind to `PRODUCER` unless it synchronously waits for a response: then it should use `CLIENT`.

Author: Also this portion. It makes sense I think to have it as no matter if you are sending the event via HTTP or to a Queue, you are a PRODUCER and whoever receives it is a CONSUMER. But I wonder if this will again cause confusion in the sense that people will mix CloudEvents with messaging systems?

> A producer of an event should set the span kind to `PRODUCER` unless it synchronously waits for a response: then it should use `CLIENT`.

Reviewer 2: ```suggestion
A emitter of an event should set the span kind to `PRODUCER`.
```

I think waiting for a response event is very unusual, and would probably be handled in another span.

Author: Yeah that makes sense. I used `sender` and `receiver` throughout the document, as it's the wording used in the CloudEvents spec. https://github.com/cloudevents/spec/blob/master/primer.md
